hazelcast/src/main/java/com/hazelcast/partition/impl/InternalPartitionServiceImpl.java

```diff
@@ -2130,15 +2124,18 @@ private void doRun() throws InterruptedException {
         }
 
         boolean processTask(Runnable r) {
-            if (r == null || isInterrupted()) {
-                return false;
-            }
-            migrating = (r instanceof MigrateTask);
             try {
+                if (r == null || isInterrupted()) {
+                    return false;
+                }
+
                 r.run();
             } catch (Throwable t) {
                 logger.warning(t);
+            } finally {
+                migrationQueue.afterTaskCompletion(r);
```

Comment:
It does. Now you have a mo ecomplex method than needs to be. With the wrapper task approach I have given, this is fully an implementation detail of the MigrationQueue. The logic for this thread is simplified since it doesn't need to do any callbacks to the queue. In my book this is cleaner code since incrementing and decrementing the counter is a implementation detail of the MigrationQueue.